extension/background.js

```diff
@@ -0,0 +1,230 @@
+    // `redirectUrl` is both valid and different than the original URL. we
+    // likely detected a search server-side redirect.
+    const isServerSideRedirect =
+      !addonId && typeof redirectUrl !== "undefined" && url !== redirectUrl;
+
+    // Search server-side redirects are possible because an extension has
+    // registered a search engine, which is why we can (hopefully) retrieve the
+    // add-on ID.
+    if (isServerSideRedirect) {
+      const id = this.getAddonIdFromUrl(url);
+
+      // We shouldn't report built-in search engines.
+      if (!id.endsWith("@search.mozilla.org")) {
+        addonId = id;
+      }
+    }
+
+    if (!addonId) {
+      // No add-on ID means there is nothing we can report.
+      return;
+    }
+
+    // This is the (initial) URL before the redirect.
+    const from = await browser.addonsSearchExperiment.getPublicSuffix(url);
+
+    // This is the URL after the redirect.
+    const requestUrl =
+      redirectUrl ||
+      (await browser.addonsSearchExperiment.getRequestUrl(requestId));
+    const to = await browser.addonsSearchExperiment.getPublicSuffix(requestUrl);
+
+    if (from === to) {
+      // We do not report redirects to same public suffixes.
+      return;
+    }
+
+    // Hopefully this is "all" we need to prevent multiple Telemetry event
+    // submissions, which might happen when both `onBeforeRequest` and
+    // `onBeforeRedirect` listeners are called for the same request (which is
+    // possible depending on the order of the registered listeners).
+    this.lastRequestIdReported = requestId;
+
+    const telemetryObject = isServerSideRedirect
+      ? TELEMETRY_OBJECT_OTHER
+      : TELEMETRY_OBJECT_WEBREQUEST;
+
+    const telemetryValue = isServerSideRedirect
+      ? TELEMETRY_VALUE_SERVER
+      : TELEMETRY_VALUE_EXTENSION;
+
+    // Get the add-on details we need to send as extra props.
+    const addon = await browser.addonsSearchExperiment.getAddonById(addonId);
+
+    const telemetryExtra = {
+      addonId,
+      addonVersion: addon.version,
+      from,
+      to,
+    };
+
+    this.recordEvent(
+      TELEMETRY_METHOD_ETLD_CHANGE,
+      telemetryObject,
+      telemetryValue,
+      telemetryExtra
+    );
+  };
+
+  recordEvent(method, object, value, extra) {
+    console.debug(
+      `recording event: method=${method} object=${object} value=${value} extra=${JSON.stringify(
+        extra
+      )}`
+    );
+
+    browser.telemetry.recordEvent(
+      TELEMETRY_CATEGORY,
+      method,
+      object,
+      value,
+      extra
+    );
+  }
+
+  getAddonIdFromUrl(url) {
+    for (const pattern of Object.keys(this.matchPatternsMap)) {
+      const [urlPrefix] = pattern.split("*");
+
+      if (url.startsWith(urlPrefix)) {
+        return this.matchPatternsMap[pattern];
+      }
+    }
+
+    return null;
+  }
+}
+
+const start = async () => {
+  const exp = new AddonsSearchExperiment();
+  await exp.monitor();
```

Comment:
I don't think it's necessary to await here (or in the later call), so the `start` function could also be dropped.

Reply:
mm what if the listener below is called before the first `monitor()` call ends?